### PR TITLE
MAINT: optimize._chandrupatla: refactor for code reuse

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Followup to gh-17719

#### What does this implement/fix?
This refactors `optimize._chandrupatla` so that elements of the code can be reused by other vectorized scalar optimization algorithms. All the code specific to Chandrupatla's rootfinder is preserved in `_chandrupatla`, and (almost) all the algorithm-independent parts are factored out into more general functions.

The benefit is that other scalar minimization, rootfinding, and bracket-finding algorithms (maybe univariate integration like gh-18650?) can rely on common code to take care of details that are often handled inconsistently (dtype, function evaluation and iteration count, callback interface). Perhaps more importantly, takes care of an important bookkeeping task that avoids unnecessary work: eliminating elements that have already terminated. I've seen 300x speed improvements for vectorized rootfinding with NumPy arrays (https://github.com/scipy/scipy/pull/17719#issuecomment-1547778146), and this can easily be extended to take advantage of GPU arrays types once that infrastructure matures (see gh-18668).

#### Additional information
Todo:
- [x] Factor out algorithm-independent parts of `prepare_result`
- [x] Improve documentation; naming
- [ ] (After review) Move algorithm-independent parts to separate file
- [ ] ~~Add generic, property-based tests~~ (separate PR)

### CODE REVIEW

File: scipy/optimize/_zeros_py.py [hunk 1]
```diff
@@ -1514,37 +1514,191 @@ def _chandrupatla(func, a, b, *, args=(), xatol=_xtol, xrtol=_rtol,
...
+    - It handles common tasks involving iteration count, function evaluation
+      count, a user-specified callback, and associated termination conditions.
+    - It compresses the attributes of `work` to eliminate unnecessary
+      computation on elements that have already converged.
+
+    """
+    cb_terminate = False
+
+    # Initialize the result object and active element index array
+    n_elements = int(np.product(shape)) or 1
```
Comment (mdhaber) [inline scipy/optimize/_zeros_py.py:1682]:
```suggestion
    n_elements = int(np.prod(shape)) or 1
```

**Comment by mdhaber** [comment]
@tupui Once this is merged, it will be easy to add:

- `_bracket_root` - a refactor of https://github.com/scipy/scipy/pull/18348 (needed for the inverse CDF/CCDF and `fit` methods of new distributions)
- `_bracket_minimum` - a vectorized scalar minimization bracket-finder (needed for the `mode` and `fit` methods of new distributions)
- `_minimize_chandrupatla` - Chandrupatla's minimization algorithm (needed for the `mode` and `fit` methods of new distributions)
- `_integrate` - An upgrade of https://github.com/scipy/scipy/pull/18650 (needed for the CDF/CCDF methods of new distributions)
- `_differentiate` - Adaptive numerical differentiation (needed for `fit` methods of new distributions).

To show that this framework is pretty flexible already (and of course can be tweaked as needed), I already took a shot at `_differentiate` in https://github.com/mdhaber/scipy/pull/106. It's worth taking a look to see how new functions using this framework can focus on the algorithmic parts; all the rest comes for free.

**Comment by mdhaber** [comment]
The last commit provides a nice speed boost.

```python3
import numpy as np
from scipy import stats
from scipy.optimize._zeros_py import _chandrupatla

dist = stats.argus(1)
ps = np.linspace(0.005, 0.995, 100000)
# %timeit -n1 -r1 [optimize.brentq(lambda x: dist.cdf(x) - p, 0.001, 0.999) for p in ps]
# 1min 30s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)
q = [optimize.brentq(lambda x: dist.cdf(x) - p, 0.001, 0.999) for p in ps]
fmax = np.max(np.abs(dist.cdf(q)-ps))
assert fmax > 1e-15

# being careful to achieve even better accuracy
# %timeit _chandrupatla(lambda x, ps: dist.cdf(x) - ps, 0.001, 0.999,
#                       xatol=0, xrtol=0, fatol=1e-15, args=(ps,))
# 203 ms ± 4.41 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

res = _chandrupatla(lambda x, ps: dist.cdf(x) - ps, 0.001, 0.999,
                    xatol=0, xrtol=0, fatol=1e-15, args=(ps,))
fmax2 = np.max(np.abs(dist.cdf(res.x)-ps))
assert fmax2 <= 1e-15
```
So that shaved about 20% off the execution time compared to https://github.com/scipy/scipy/pull/17719#issuecomment-1547778146, and it's now more than 440x times faster than `brentq` here.

**Review by tupui — APPROVED** [review]
Thanks @mdhaber. I pulled it locally and played with it, I could confirm the speedup. LGTM 🚀 

I think we should continue and get the ball rolling (whole CI agrees too.) Using this framework with other methods will help to polish the design (and we don't want to do this in this PR.)

For now the design seems sensible. There is an argument to use a more OOP approach: maybe like the optim class or even subclass or make some factory, also use more data structures instead of long tuples. But this is also fine like that. We can see later as the bulk of the work is going to be the same, having a clear separation of responsibilities like that will allow us to be flexible about the structure.

Putting this here. What we need to do after this work settle is maybe to include a flow diagram and a few explainers like `active` and compression (I think the concepts are still blurry just reading the doc parts. Maybe providing a tiny e.g. would be helpful. I am trying to help the future us.)

Let's do all the moves in a follow up. No need to pile up here, we can iterate with small PRs too. For the move, if we say we move Chandrupatla, I suppose we will also want in the end to move the other methods as well. So I would suggest to do all moves at once. Then we would use the new framework in separate PRs.

**Pull request merged**